File: src/plot.rs
use crate::*;
use std::io::Write;

pub const COLOR_A: &str = "#1f77b4";
pub const COLOR_M: &str = "#ff7f0e";
pub const COLOR_E: &str = "#2ca02c";

static ID2NAME: [&str; 3] = ["e", "n", "u"];

pub struct Plot<'a> {
    plot: plotly::Plot<'a, std::fs::File, &'static str>,
    nrows: usize,
    row_ids: std::collections::HashMap<String, usize>,
    trace_prefix: Option<String>,
}

impl<'a> Plot<'a> {
    pub fn new<P: AsRef<std::path::Path>>(path: P) -> Result<Self, Error> {
        let mut f = std::fs::File::create(path)?;

        f.write_all(b"<html><head><meta charset=\"utf-8\"/>")?;
        f.write_all(b"<style>.modebar-container {position: sticky !important;} .infolayer text[class^=\"y\"][class$=\"title\"] {transform: rotate(0deg); writing-mode: vertical-lr; text-orientation: upright; font-weight: bold !important;}</style>")?;
        write!(
            &mut f,
            "<script src=\"{}\" charset=\"utf-8\"></script>",
            plotly::URL_CDN
        )?;
        f.write_all(b"</head><body style=\"margin:0; padding:0; overflow-x:hidden\">")?;
        f.write_all(b"<script type=\"text/javascript\">")?;

        let mut plot = Self {
            plot: plotly::Plot::new(f, "plotly-div")?,
            nrows: 0,
            row_ids: std::collections::HashMap::new(),
            trace_prefix: None,
        };
        let config = plot.plot.config();

        config
            .responsive(true)
            .display_mode_bar(plotly::config::DisplayModeBar::True);
        config.scroll_zoom().set(plotly::config::ScrollZoom::True);

        plot.plot
            .layout()
            .legend()
            .traceorder()
            .flags()
            .grouped(true);

        Ok(plot)
    }

    pub fn set_trace_prefix<A: AsRef<str>>(&mut self, trace_prefix: Option<A>) {
        self.trace_prefix = trace_prefix.map(|v| v.as_ref().to_string());
    }

    pub fn add_row<A: AsRef<str>>(&mut self, name: Option<A>) -> Result<usize, Error> {
        if let Some(name) = name {
            if self.row_ids.get(name.as_ref()).is_some() {
                return Err(Error::RowAlreadyExists);
            }

            self.row_ids.insert(name.as_ref().to_string(), self.nrows);
            self.plot
                .layout()
                .yaxis(self.nrows)
                .title()
                .text(name.as_ref().to_string());
        }

        self.nrows += 1;

        Ok(self.nrows - 1)
    }

    pub fn add_trace(&mut self, trace: &mut plotly::traces::scatter::Scatter) -> Result<(), Error> {
        if self.nrows == 0 {
            return Err(Error::NoRow);
        }

        self.add_trace_to_rowid(trace, self.nrows - 1)
    }

    pub fn add_trace_to_rowname<A: AsRef<str>>(
        &mut self,
        trace: &mut plotly::traces::scatter::Scatter,
        name: A,
    ) -> Result<(), Error> {
        if let Some(id) = self.row_ids.get(name.as_ref()) {
            let id = *id;
            self.add_trace_to_rowid(trace, id)
        } else {
            Err(Error::RowNotFound)
        }
    }

    pub fn add_trace_to_rowid(
        &mut self,
        trace: &mut plotly::traces::scatter::Scatter,
        id: usize,
    ) -> Result<(), Error> {
        if id >= self.nrows {
            return Err(Error::RowNotFound);
        }

        trace.yaxis(format!("y{}", id + 1));
        trace.xaxis(format!("x{}", id + 1));

        if let Some(trace_prefix) = &self.trace_prefix {
            trace.name(format!(
                "{}{}",
                trace_prefix,
                trace.name.as_ref().map(|v| v.as_ref()).unwrap_or("")
            ));
        }

        self.plot.add_trace(trace)?;
        Ok(())
    }

    pub fn finish(mut self) -> Result<(), Error> {
        self.plot.layout().grid().rows(self.nrows as u64).columns(1);

        let mut f = self.plot.finish()?;

        let height = (100.0f64 / 3.0 * self.nrows as f64).max(100.0);
        f.write_all(b"</script>")?;
        write!(
            &mut f,
            "<div id=\"plotly-div\" style=\"width:100%;height:{}vh;\"></div>",
            height
        )?;
        f.write_all(b"<script type=\"text/javascript\">")?;
        f.write_all(include_bytes!("js/plot.js"))?;
        f.write_all(b"</script>")?;
        f.write_all(b"<script type=\"text/javascript\">makeplot();</script>")?;
        f.write_all(b"</body></html>")?;

        Ok(())
    }

    pub fn default_line<'b>() -> plotly::traces::scatter::Scatter<'b> {
        let mut t = plotly::traces::scatter::Scatter::default();
        t.line().simplify(false);
        t
    }

    pub fn axisid_to_rowname(name: &str, id: usize) -> String {
        format!("{}-{}", name, ID2NAME[id])
    }

    pub fn add_measurements(&mut self, samples: &[Data], x: &[f64]) -> Result<(), Error> {
        let mut add_row = |rowname: &str, id: usize, y: &[f64]| -> Result<(), Error> {
            let mut t = Self::default_line();
            t.x(&x).y(&y).name("measurement");
            t.line().color(COLOR_M);

            self.add_row(Some(Self::axisid_to_rowname(rowname, id)))?;
            self.add_trace(&mut t)?;

            Ok(())
        };

        for i in 0..3 {
            let y: Vec<f64> = samples.iter().map(|s| s.accel[i]).collect();
            add_row("acc", i, &y)?;
        }

        for i in 0..3 {
            let y: Vec<f64> = samples.iter().map(|s| s.gyro[i]).collect();
            add_row("gyr", i, &y)?;
        }

        for i in 0..3 {
            let y: Vec<f64> = samples.iter().map(|s| s.mag[i]).collect();
            add_row("mag", i, &y)?;
        }

        Ok(())
    }
}

File: src/js/plot.js
function axis2id(name) {
    if (name == "y") {
        return 0;
    }

    return parseInt(name.substring(1), 10) - 1;
}

function id2axis(id) {
    if (id == 0) {
        return "y";
    }

    return "y" + (id + 1);
}

function id2layoutaxis(id) {
    if (id == 0) {
        return "yaxis";
    }

    return "yaxis" + (id + 1);
}


var groups = [];
var rowconfig = {};
try {
    rowconfig = JSON.parse(window.localStorage.rowconfig);
}
catch(err) {
}

for (var traceid=0; traceid<data.length; traceid++) {
    var trace = data[traceid];
    const id = axis2id(trace.yaxis);
    if (!(id in groups)) {
        groups[id] = {
            "name": null,
            "traces": [],
        };

        const axisname = id2layoutaxis(id);
        if (axisname in layout && "title" in layout[axisname] && "text" in layout[axisname].title) {
            groups[id].name = layout[axisname].title.text;
        }
    }

    groups[id].traces.push(trace);
}

var plotsel = document.createElement("div");
plotsel.style.position = "fixed";
plotsel.style.width = "100%";
plotsel.style.height = "100%";
plotsel.style.top = 0;
plotsel.style.left = 0;
plotsel.style.background = "white";
plotsel.style.display = "none";
document.body.appendChild(plotsel);

var form = document.createElement("form");
plotsel.appendChild(form);

var list = document.createElement("div");
list.style.width = "100%";
list.style.height = "90%";
list.style.overflow = "auto";
form.appendChild(list);

for (var groupid=0; groupid<groups.length; groupid++) {
    var group = groups[groupid];

    var input = document.createElement("input");
    input.setAttribute("type", "checkbox");
    input.setAttribute("id", group.name);
    input.setAttribute("name", group.name);
    input.checked = true;
    list.appendChild(input);

    var label = document.createElement("label");
    label.setAttribute("for", group.name);
    label.innerText = group.name;
    list.appendChild(label);

    list.appendChild(document.createElement("br"));
}

var submit = document.createElement("input");
submit.setAttribute("type", "submit");
submit.setAttribute("value", "save");
submit.style.width = "100%";
submit.style.height = "10%";
submit.style.border = "none";
submit.style.background = "#00000011";
submit.style.cursor = "pointer";
form.appendChild(submit);

form.onsubmit = function(e) {
    for (var groupid=0; groupid<groups.length; groupid++) {
        var group = groups[groupid];
        var checked = list.querySelector("#"+group.name).checked;
        rowconfig[group.name] = checked;
    }

    window.localStorage.rowconfig = JSON.stringify(rowconfig);
    update_plot(true);

    plotsel.style.display = "none";
    document.body.style.overflowY = "auto";
    return false;
};

config["modeBarButtonsToAdd"] = [
    {
        name: 'plotsel',
        click: function(gd) {
            for (var groupid=0; groupid<groups.length; groupid++) {
                var group = groups[groupid];
                var checked = true;
                if (group.name in rowconfig && rowconfig[group.name]===false) {
                    checked = false;
                }
                list.querySelector("#"+group.name).checked = checked;
            }

            plotsel.style.display = "block";
            document.body.style.overflowY = "hidden";
        }
    },
];

function update_plot(redraw) {
    data.splice(0, data.length);
    var rowid = 0;
    for (var groupid=0; groupid<groups.length; groupid++) {
        var group = groups[groupid];

        if (group.name !== null && (group.name in rowconfig) && rowconfig[group.name]===false) {
            continue;
        }

        for (var traceid=0; traceid<group.traces.length; traceid++) {
            var trace = group.traces[traceid];
            trace.yaxis = id2axis(rowid);

            data.push(trace);
        }

        layout[id2layoutaxis(rowid)].title = group.name;

        rowid++;
    }

    layout.grid.rows = rowid;
    document.getElementById("plotly-div").style.height = Math.max(100.0 / 3.0 * rowid, 100.0) + "vh";

    if (redraw === true) {
        makeplot();
    }
}

update_plot(false);